content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
+
+![phase1](total-runtime.jpg)
+
+For all queries generated, the compilation version (once compiled) greatly outperforms Postgres or matches it in the case of Q11. I suspect that the noise of measurement outweights any improvement in runtime given how fast both systems execute the query. However, once we take into account the compilation time, a different result emerges.
+
+### Compilation Time
+
+![phase1](compile.jpg)
+
+At such low table sizes, the time spent compiling actually dominates. Note that the time spent compiling is fixed so as the database size grows larger, the yellow portion of the graph will increase and the grey portion will stay constant. This means the ratio of time spent compiling to time spent executing will decrease on larger databases.
+
+However, this is still unacceptable on smaller data sets where Postgres outperforms the compilation version. In future work, I discuss some ways that this can be fixed.
+
+## Personal Experiences
+
+The largest barrier I had to overcome in working on this project was understanding all the pieces of the puzzle. I needed to understand each portion of the system enough to implement it.
+
+## Future Work
+
+While this generates extremely high performant code, the time spent compiling is extremely large. The following strategies are useful in reducing this:
+- Directly generate LLVM. This removes any overhead for file I/O to output the C++ and parsing, type checking and converting that C++ to LLVM.
+- Control which optimization passes run on this. For example, most of the overhead comes from dead copies that get inserted into the code. One GVN/DCE pass can get rid of these. Auto-vectorization would also prove beneficial here.
+
+As always, there are a myriad of SQL features that I have not implemented as part of this project. Adding more features would allow me to test this more thoroughly.
+
+Another aspect to explore is how best to parallelize the generate code. Existing techniques break the code up into the materialization boundaries and then use simple strategies to parallelize it.
+
+Finally, to simplify the codebase, we can take advantage of operator overloading to define proxy types. For example, if you have `int`s a and b, writing a * b in your source code generates code that multiplies a and b. One more layer of indirection will simplify this. For example, if you have `proxy[int]`s a and b, writing a * b in your source code generates code that generates a program that multiplies a and b. This technique allows you to write code that behaves like the compilation version but appears like the interpreted version.
+
+## Sources
+
+I used the following sources as references:
+- [NoisePage](noise.page)
+- https://www.vldb.org/pvldb/vol4/p539-neumann.pdf
+- https://db.in.tum.de/~kersten/Tidy%20Tuples%20and%20Flying%20Start%20Fast%20Compilation%20and%20Fast%20Execution%20of%20Relational%20Queries%20in%20Umbra.pdf?lang=en
+- https://www.cs.purdue.edu/homes/rompf/papers/tahboub-sigmod18.pdf
```

Comment:
Rather than just listing URLs, it would be more helpful to link to this stuff in context. Like, what did you get out of learning about NoisePage, and what part did it help you with? Link to it from that section by saying "this idea was inspired by NoisePage" or something.

If there is no more context to be given, and you really do want some miscellaneous context-free bullet points at the bottom of your post, at least use actual citations when you're referring to a paper—i.e., give the title and authors instead of just the URL.

Reply:
Integrated sources like you mentioned.